### PR TITLE
[#256] Add possibility to specify fallback directly from ServiceRestClient when using Hystrix

### PR DESCRIPTION
**Pull request by kurlenda**

### CODE REVIEW

**Comment by marcingrzejszczak** [comment]
Thanks for the PR :) https://travis-ci.org/4finance/micro-infra-spring/builds/49811396, https://snap-ci.com/4finance/micro-infra-spring/pull/257/logs/defaultPipeline/1/BOOT-ACCEPTANCE You've broken the boot-microservice with your commit I guess :P

File: micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/request/HttpMethod.groovy [hunk 1]
```diff
@@ -56,4 +56,14 @@ interface HttpMethod<U, T> {
...
+    /**
+     * Adds Hystrix circuit breaker with fallback around every REST call.
+     *
+     * Example:
+     * <code>
+     * .withCircuitBreaker(HystrixCommand.Setter.withGroupKey(HystrixCommandGroupKey.Factory.asKey("Group"))
+     *      .andCommandKey(HystrixCommandKey.Factory.asKey("Command")), {return new ResponseEntity<String>("service unavailable", HttpStatus.METHOD_FAILURE)})
+     * </code>
+     */
+    HttpMethod<U, T> withCircuitBreaker(HystrixCommand.Setter setter, Closure closure)
```
Comment (marcingrzejszczak) [inline micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/request/HttpMethod.groovy]: Please add a Javadoc for the Closure element

Comment (nurkiewicz) [inline micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/request/HttpMethod.groovy]: Rename `closure` to `fallback` or something.

File: micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/response/receive/MethodParamsApplier.groovy [hunk 2]
```diff
@@ -71,4 +71,10 @@ trait MethodParamsApplier<M, EM, PM> implements HttpMethod<M, PM>, HttpEntitySen
         params.hystrix = setter
         return this
     }
+
+    @Override
+    HttpMethod<M, PM> withCircuitBreaker(HystrixCommand.Setter setter, Closure closure) {
```
Comment (nurkiewicz) [inline micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/response/receive/MethodParamsApplier.groovy]: Rename `closure` to `fallback` or something.

**Comment by szpak** [comment]
> You've broken the boot-microservice with your commit I guess :P

I hoped the same (in a positive way), but unfortunately it is only a problem with our new acceptance tests in Geb.
```
StaleElementReferenceException: Element is no longer attached to the DOM
```
(especially that the same test passed on Travis)

I will rerun it.

**Comment by szpak** [comment]
I missed the next build. Now it fails on Travis... We need to talk with @thiefik on Monday.

**Comment by thiefik** [comment]
I'm investigating this, but locally everything is working fine. Probably in this Geb test is some race condition.

**Comment by thiefik** [comment]
Improved GEB tests in https://github.com/4finance/boot-microservice:
https://github.com/4finance/boot-microservice/commit/8c8bd1e1f81baa56f01110b61fce9f982ea0605f
https://github.com/4finance/boot-microservice/commit/a4e0cbd1ba6a42e823d0f8da7cac04523dcd93b2

File: micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/response/executor/RestExecutor.groovy [hunk 3]
```diff
@@ -80,25 +80,38 @@ final class RestExecutor<T> {
...
+                        return httpInvocation.call()
+                    }
+
+                    @Override
+                    protected ResponseEntity<T> getFallback() {
+                        return hystrixFallback.call()
+                    }
+                }.execute()
+            }
             return new CorrelatedCommand<ResponseEntity<T>>(hystrix) {
```
Comment (mchmielarz) [inline micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/response/executor/RestExecutor.groovy]: It's rather cosmetic one but could we use `else` here to highlight the second branch of if-else statement?

**Comment by mchmielarz** [comment]
LGTM

File: micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/response/executor/RestExecutor.groovy (hunk 3 above)
Comment (marcingrzejszczak) [inline micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/resttemplate/fluent/common/response/executor/RestExecutor.groovy]: Personally I like such statements :) I would keep it that way but it's up to you @kurlenda :)

**Comment by marcingrzejszczak** [comment]
IMO if we resolve @mchmielarz doubt then we could merge that - this PR is really cool :)

**Pull request merged**